### PR TITLE
Concurrent B+ Tree

### PR DESCRIPTION
**Pull request by abhijithanilkumar**
# B+ Tree for NoisePage

## Description

This PR introduces a concurrent B+ Tree that can potentially replace the existing BwTree implementation. The implementation uses latch crabbing, with per-node shared latches for synchronization. For reads, shared latches are acquired throughout the path. For writes, exclusive latches are acquired, and all parent latches are released if the current node is safe (no overflow/underflow possible).

There are two types of nodes:
* InnerNode: Non-leaf nodes in the Tree. Stores a key and the pointer to its child node. The node itself acts as a vector of `<key, ptr>` pairs. This vector is manually managed by allocating and deallocating memory whenever necessary.
* LeafNode: Leaves of the Tree, storing actual values. Stores a key and a list of values for each key. The list of values is currently implemented as `std::unordered_set`~`std::list` (An implicit drawback of this is that the lookups are O(n), can be improved to O(logn) when there are multiple values in the same key. Can be potentially improved further by switching to an unordered_set, since ordering is not important for these values)~.

`ScanKey`, `ScanAscending`, `ScanDescending` and `ScanLimitDescending` are supported. The scans will abort and retry if the latch on a particular node cannot be acquired (in order to make sure that this acquisition pattern does not deadlock with concurrent scans and inserts/deletes).

The initial code for the B+ Tree has been ported over from 15721 Project-2 (Spring 2020), implemented by @rohanaggarwal7997 @gautam20197 and @preetansh .

## Remaining tasks

- [x] Use binary search to search among values in the LeafNode (potentially switch to unordered_set)

EDIT: Even though `std::unordered_set` improves performance of the system, it takes up more memory. Experiments suggest that `std::list` uses less memory while taking up slightly more time. Therefore, switched back to `std::list`.

## Performance

Will be updating this section with the actual time and memory footprint after running OLTPBench on the B+ tree.

### CODE REVIEW

**Comment by abhijithanilkumar** [comment]
Review Request:
* `src/include/storage/index/bplustree.h` : 
  * Tree Structure (1-910) - @thepinetree @kjobanputra [Since both of reviewers will need background regarding tree structure]
  * Insert Functions + ScanAscending ScanDescending + ScanLimitDescending - @kjobanputra (1326-1537, 1540-2188)
  * Delete Functions + ScanAscending - @thepinetree (1200-1312, 2196-2498)
  * Helper functions - @thepinetree (2511-2552)
  * Structural Integrity Verification - @kjobanputra (921-1184) [This code is specific to tests, verifying structural integrity]
  
* `test/storage/bplustree_test.cpp` : 
  * 1-800 - @thepinetree 
  * 801-1468 - @kjobanputra 

* Remaining Files: @turingcompl33t [The remaining files are a direct replacement to the equivalent bwtree files, so should be easy to review] 
  * `bplustree_benchmark` and `bplustree_index_test` are basically the same as `bwtree_benchmark` and `bwtree_index_test`, with function name changes specific to B+ Tree.
  * `bplustree_index.*` is analogous to `bwtree_index.*`. The major change is that scan functions in B+ Tree has a retry in case it fails because a lock cannot be acquired.
  * The remaining changes are pretty minor.

**Review by turingcompl33t — COMMENTED** [review]
None of the comments in this review should block merging this PR, except perhaps the question raised in the implementation of `BPlusTreeIndex::ScanKey()` regarding duplicate keys spread across multiple leaf nodes. Will explicitly approve (for my files at least) once that question is addressed.

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:111]: Can you write something more specifically about why the root pointer needs two latches for safety? It's not clear to me which case only having one latch for the root is unsafe.

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:490]: nit: typo (next instead of exnt)

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [reply src/include/storage/index/bplustree.h:111]: It's in the case where the root is being replaced, right? If so, it'd be nice to make that clearer in the comment.

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:1646]: Maybe when talking about the structure of the tree state that we allow duplicate keys

File: src/include/storage/index/bplustree.h
Comment (mush-zhang) [inline src/include/storage/index/bplustree.h:802]: should we use `KeyCmpEqual` here?

File: src/include/storage/index/bplustree.h
Comment (mush-zhang) [inline src/include/storage/index/bplustree.h:1174]: Misleading documentation copied from FreeTree?

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:1069]: Style nit: Keep in loop guard

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:1108]: Same small bit: keep in loop guard

File: src/include/storage/index/bplustree.h
Comment (kjobanputra) [inline src/include/storage/index/bplustree.h:1138]: Is there a reason that a coarse lock is used?

**Review by mush-zhang — CHANGES_REQUESTED** [review]
Generally LGTM.Very clear code and documentation. Thank you! I have left some minor comments and questions. You could probably add some testcases covering those edge cases pointed out by codecov. Another thing: Do we want the test functions in the same bplus tree header file? we could create a TestBplusTree class for those functions and put them into a separate file.

File: test/storage/bplustree_test.cpp
Comment (kjobanputra) [inline test/storage/bplustree_test.cpp:1218]: How is this different than the test above?

File: test/storage/bplustree_test.cpp
Comment (kjobanputra) [inline test/storage/bplustree_test.cpp:1401]: Also include the structural integrity check

File: test/storage/bplustree_test.cpp
Comment (kjobanputra) [inline test/storage/bplustree_test.cpp:1456]: Same comment as above: add structural integrity check

**Review by kjobanputra — CHANGES_REQUESTED** [review]
My comments are mostly nitpicking: some commenting changes, a question about a design decision, and including calls to check the structural integrity of the tree in the later tests. Once those are addressed, I'll approve. Good job overall!

File: src/include/storage/index/bplustree.h
Comment (tpan496) [inline src/include/storage/index/bplustree.h:220]: Would it be better to comment as "The depth of the node within the tree" instead of using height?

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:31]: Would it be better to create local variables for those magic numbers to improve readability?

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:77]: Similar comment to the one above. I think it would be better to name "i" to something more meaningful.

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:25]: Maybe some brief comments on what each test does?

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:519]: Similar style comments apply here

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:677]: Can we comment on why this is considered a "Larger" test? The numbers do seem large, but I think it would be more clear to clarify this.

File: test/storage/bplustree_test.cpp
Comment (tpan496) [inline test/storage/bplustree_test.cpp:1177]: Can we name it in a more meaningful way instead of just adding 2 at the end?

File: benchmark/storage/bplustree_benchmark.cpp
Comment (abhijithanilkumar) [reply benchmark/storage/bplustree_benchmark.cpp]: Done

File: benchmark/storage/bplustree_benchmark.cpp
Comment (abhijithanilkumar) [reply benchmark/storage/bplustree_benchmark.cpp]: I believe all the microbenchmarks use this division as of now, we can probably make the change to ScopedTimer and change it everywhere, but I'm not sure if that is worth the effort.

File: src/include/storage/index/bplustree_index.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree_index.h:87]: I feel you! This interface will be removed in the near future when we can stop materializing keys here. Instead use an index iterator in the execution engine.

File: src/include/storage/index/bplustree_index.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree_index.h:43]: This function is implemented in the generic index wrapper and a change here will have a bigger blast radius. This will need changes to the corresponding implementations in BwTree and Hash index. I'll keep this in mind and make the change in the next PR removing the BwTree.

File: src/include/storage/index/bplustree_index.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree_index.h:57]: This change will require modification to all the index implementations. We will probably need a broader discussion on whether we should make this change.

File: src/include/storage/index/index_builder.h
Comment (abhijithanilkumar) [reply src/include/storage/index/index_builder.h]: I've made the change for B+ Tree, not changing the other index implementations, we can probably do this in a later PR.

File: src/storage/index/bplustree_index.cpp
Comment (abhijithanilkumar) [reply src/storage/index/bplustree_index.cpp]: Done

File: src/storage/index/bplustree_index.cpp
Comment (abhijithanilkumar) [reply src/storage/index/bplustree_index.cpp:103]: It's not possible to have multiple keys with the same value at the leaf level. Each key can hold multiple values though. So, when we scan for a key, it is essentially a point lookup and we are guaranteed that the key that pops out of the search is the only one present.

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h:1139]: This function returns the **exact** heap usage. So technically, if we use fine-grained locking, the measurement will not be accurate since the tree can change dynamically. This function sort of captures the current snapshot. For estimates of heap usage, EstimateHeapUsage() function should be called.

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h:152]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h:523]: This is not a `PushBack` because this is to insert an element at a specific location while `PushBack` always inserts it at the end.

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h:602]: This is a check to verify that the index is in bounds. There are some parts of delete code that relies on this check.

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Done

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h]: Modified

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:1218]: This test deletes elements in the reverse order. Will include some meaningful documentation on what each test does while addressing @tpan496's comments below.

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:1401]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:1462]: Done

File: src/storage/index/bplustree_index.cpp
Comment (turingcompl33t) [reply src/storage/index/bplustree_index.cpp:103]: Great, thanks for the clarification!

**Review by turingcompl33t — APPROVED** [review]
All of the points in my initial review were addressed. Great work @abhijithanilkumar!

**Review by kjobanputra — APPROVED** [review]

**Review by mush-zhang — APPROVED** [review]
LGTM

File: src/include/storage/index/bplustree.h
Comment (abhijithanilkumar) [reply src/include/storage/index/bplustree.h:223]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:38]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:588]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp:766]: Done

File: test/storage/bplustree_test.cpp
Comment (abhijithanilkumar) [reply test/storage/bplustree_test.cpp]: Removed these 2 tests since they were redundant.

**Comment by tpan496** [comment]
LGTM

**Review by lmwnshn — APPROVED** [review]
Rubber stamping, see `#pr-czar` in slack.

**Pull request merged**
